barretenberg/cpp/src/barretenberg/vm/avm/trace/trace.cpp

```diff
@@ -4599,6 +4737,16 @@ std::vector<Row> AvmTraceBuilder::finalize()
 
     gas_trace_builder.finalize(main_trace);
 
+    if (apply_end_gas_assertions) {
+        // Sanity check that the amount of gas consumed matches what we expect from the public inputs
+        auto last_l2_gas_remaining = main_trace.back().main_l2_gas_remaining;
+        auto expected_end_gas_l2 = public_inputs.gas_settings.gas_limits.l2_gas - public_inputs.end_gas_used.l2_gas;
+        ASSERT(last_l2_gas_remaining == expected_end_gas_l2);
+        auto last_da_gas_remaining = main_trace.back().main_da_gas_remaining;
+        auto expected_end_gas_da = public_inputs.gas_settings.gas_limits.da_gas - public_inputs.end_gas_used.da_gas;
+        ASSERT(last_da_gas_remaining == expected_end_gas_da);
+    }
+
```

Comment:
Need this boolean flag otherwise we need the right end gas in every test which is too much of a pain.